### PR TITLE
Removes some major sleepers, Refactors maploading to be async (100% lag free) 4

### LINKED ISSUE #7927: Pubby Mining Shuttle disappearing and deleting miners on Pubby
## Round ID & Round Date: 
40898 on 17/10/2022

## Testmerges:
#7430
#7659
#7500

## Reproduction:

On Pubby, the mining shuttle spawns missing the entire floor and a few walls. (Screenshot is after attempted repairs and launching the shuttle)
image
(Screenshot after landing - the miner got deleted)
image
Upon further inspection, it looks like all of the other shuttles (apart from the exploration and labour shuttles) have suffered from the same problem.
image

### LINKED ISSUE #7967: Shuttles spawning horribly broken.
## Round ID & Round Date:
Round ID: 41032
October 2022 25th

## Testmerges:

#7659: 'Removes some major sleepers (100% lag free)
#7500: '[TG PORT] Space Dragon Rework'

## Reproduction:

Unsure if can reproduced reliably. I've seen it happen I think three times now on low-pop, unsure on others. Sometimes, it seems shuttles, not all but sometimes a handful, bug out and decide not to fully spawn, instead spawning partially, with dirt and grime just floating on space tiles. When people use these shuttles they get launched into hyperspace and die a slow painful death that's very hard to recover from. 
Screenshot (1248)
The only idea I have for what could be causing this would be the Sleepers Rework? I know that touched upon spawning things.

### PR DESCRIPTION
**Pull request by PowerfulBacon**
Thanks to TheFinalPotato from coderbus for helping diagnose this issue.

closes #7967
closes #7927

## Downstream Notice: Breaking changes

This PR changes the way that /datum/map_template are loaded.

Previously, map_templates would have their completion awaited for, however now the loading operates on a subsystem, meaning .load() will no longer freeze the current proc until loading is completed.
/datum/parsed_map (Used for loading the station at roundstart and loading z-levels) is not affected by this and will run using CHECK_TICK (This is fine for roundstart, but shouldn't be used during a round).
All instances of /proc/load from /datum/map_template and subtypes need to be checked to ensure that any code after .load() doesn't depend on the results of .load.
Any code that needs to be run after the completion of the map load need to use the completion callbacks.

**Completion Callbacks:**

If you are calling /datum/map_template.load() from somewhere else: Map template load returns the map generator being used to generate the map, you can use .on_completion to register a completion callback which will be called after the map successfully loads.
image

---

If you are calling /datum/map_generator.generate(), then you can register completion callbacks using the .on_completion() proc to register as many completion procs as you want.
image

---

Old code:
```dm
ship.load(T)
message_admins("Ship loaded")
```
New code:

```dm
/datum/whatever/proc/load_ship(turf/T)
	var/datum/map_generator/ship_placer = ship.load(T)
	ship_placer.on_completion(CALLBACK(src, .proc/on_ship_loaded))

/datum/whatever/proc/on_ship_loaded(datum/map_generator/map_gen, turf/T, init_atmos, datum/parsed_map/parsed, register = TRUE)
	message_admins("Ship loaded")
```

If you want, you can probably make the code closer to /TG/ by just using UNTIL and waiting until the generation is completed, however its a somewhat janky solution and won't be able to utilize all of the subsystem for concurrent loading.

## About The Pull Request

Removes some major sleepers that could result in the MC freezing up (Sleeping procs may wake up before the MC, leading to the MC's maptick being consumed before the MC has a chance to fire).

## Why It's Good For The Game

Sleeping inside some extremely long function, or something that loops can cause the sleeps to wake up before the MC, meaning the MCs maptick is consumed before it even has a chance to fire, leading to the MC freezing up and everything going to shit.

## Testing Photographs and Procedure

Round > Trigger Event > Catastrophic Meteors
Watch as meteors approach the station and blow shit up.
Some systems will queue up, however the MC will still run and that list will slowly get shorter and shorter.
The MC no longer freezes during meteors as a result of airlocks losing power.

Ruin Generation:
Harder to test this one since it never happened locally anyway. Just try it on the server and hope for the best.
Of course I tested to make sure the ruins generate properly. I'm going to TM this anyway.

I've tested every form of map place that is affected, except for fugatives since it requires players to trigger the event.

image

image

image

## Changelog
:cl: PowerfulBacon, TheFinalPotato
fix: Removes a sleep loop from airlocks
fix: Makes ruin generation a subsystem process rather than using CHECK_TICK, so that its execution flow can be managed by the MC rather than byond.
code: Adds in a unit test for shuttle files that are just completely broken.
/:cl:

### CODE REVIEW

**Review by itsmeow — DISMISSED** [review]
This thing has been testmerged so goddamn long, I don't care WHAT the code is, it actually does seem to help performance

**Comment by PowerfulBacon** [comment]
Took 4 attempts to wrangle git, the mapdiff helper script was having some really weird interactions with it and git was refusing to accept my pushes, even when forced for some reason so the old PRs weren't in sync with what I had locally. Weird

**Review by itsmeow — APPROVED** [review]

**Comment by Penwin0** [comment]
Has been TMd forever, bacon says it's ready to merge and nothing looks glaringly wrong to me, it's going in

File: code/modules/mapping/reader.dm [hunk 1]
```diff
@@ -145,7 +155,7 @@
 	Master.StopLoadingMap()
 
 // Do not call except via load() above.
-/datum/parsed_map/proc/_load_impl(x_offset = 1, y_offset = 1, z_offset = world.maxz + 1, cropMap = FALSE, no_changeturf = FALSE, x_lower = -INFINITY, x_upper = INFINITY, y_lower = -INFINITY, y_upper = INFINITY, placeOnTop = FALSE)
+/datum/parsed_map/proc/_load_impl(x_offset = 1, y_offset = 1, z_offset = world.maxz + 1, cropMap = FALSE, no_changeturf = FALSE, x_lower = -INFINITY, x_upper = INFINITY, y_lower = -INFINITY, y_upper = INFINITY, placeOnTop = FALSE, check_tick = TRUE)
```
Comment (itsmeow) [inline code/modules/mapping/reader.dm:158]: this argument is never used

**Pull request merged**